### PR TITLE
Add Qwen3.5 MTP AutoDeploy support

### PR DESCRIPTION
**Pull request by govind-ramnarayan**
@coderabbitai summary

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [ ] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: examples/auto_deploy/model_registry/configs/qwen3.5_moe_400b_mtp.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  free_gpu_memory_fraction: 0.7
+  tokens_per_block: 32
+speculative_config:
+  decoding_type: MTP
+  max_draft_len: 1
+  mtp_eagle_one_model: true
+transforms:
+  insert_cached_causal_conv:
+    backend: triton_causal_conv
+  multi_stream_gemm:
```
Comment (govind-ramnarayan) [inline examples/auto_deploy/model_registry/configs/qwen3.5_moe_400b_mtp.yaml]: Let's see if we can enable the multi stream ops and see if the accuracy test still consistently passes?

File: examples/auto_deploy/model_registry/models.yaml [hunk 2]
```diff
@@ -339,17 +330,23 @@ models:
 - name: Qwen/Qwen3.5-397B-A17B
   config_id: qwen3_5_moe_400b
   yaml_extra: ['dashboard_default.yaml', 'world_size_8.yaml', 'qwen3.5_moe_400b.yaml']
+- name: Qwen/Qwen3.5-397B-A17B-FP8
```
Comment (govind-ramnarayan) [inline examples/auto_deploy/model_registry/models.yaml]: nit: No need to add the non-mtp FP8 version in this PR

File: tensorrt_llm/_torch/auto_deploy/custom_ops/fla/fla_backend_gated_delta.py [hunk 3]
```diff
@@ -139,14 +148,72 @@ def fla_cached_gated_delta_rule(
...
+        extend_end = extend_start + num_extend_tokens
+        slot_idx_extend = slot_idx[num_prefill : num_prefill + num_extend].contiguous()
+
+        q_ext = q_flat[extend_start:extend_end].view(num_extend, tokens_per_extend, H_k, K)
+        k_ext = k_flat[extend_start:extend_end].view(num_extend, tokens_per_extend, H_k, K)
+        v_ext = v_flat[extend_start:extend_end].view(num_extend, tokens_per_extend, HV, -1)
+        a_ext = a_flat[extend_start:extend_end].view(num_extend, tokens_per_extend, HV)
+        b_ext = b_flat[extend_start:extend_end].view(num_extend, tokens_per_extend, HV)
+
+        if batch_info.get_max_draft_len() > 0 and intermediate_delta_cache is not None:
```
Comment (govind-ramnarayan) [inline tensorrt_llm/_torch/auto_deploy/custom_ops/fla/fla_backend_gated_delta.py]: nit: I don't think we need this check on max_draft_len - extend requests are by definition speculative requests right now. Double-check with the triton SSM backends.

Comment (govind-ramnarayan) [reply tensorrt_llm/_torch/auto_deploy/custom_ops/fla/fla_backend_gated_delta.py]: The issue with this is the following. Please check the code path to verify what I am describing.

In the currrent code path, intermediate_delta_cache is always emitted as a cache argument by this op, whether speculation is on or off. Depending on whether speculation is on, the cache is managed (in which case it is bound to the KV Cache Manager's intermediate state cache) or it is not (in which case it is left unmanaged and is a torch.empty()). It is never actually "None"

(Btw, this is also true in triton_backend_mamba, the cache is just declared as Optional when it probably should not be - the Optional is just used for disabling in unit tests).

So effectively this if statement is useless - we just currently make the assumption that num_extend > 0 => speculation is on (so max draft len > 0) and that speculation always involves using the intermediate delta cache.

Please verify this logic and make the edit - this should eliminate the else statement below.

File: tensorrt_llm/_torch/auto_deploy/models/custom/modeling_eagle.py [hunk 4]
```diff
@@ -138,6 +141,34 @@ class EagleConfig(PretrainedConfig):
...
         },
+        "qwen3_5_moe_text": {
+            "load_embedding_from_target": True,
+            "load_lm_head_from_target": True,
+            "num_capture_layers": 1,
+            "normalize_target_hidden_state": True,
+            "layers_handle_final_norm": True,
+            "_checkpoint_conversion_mapping": {
+                r"^mtp\.layers\.0\.": "model.layers.",
+                r"^mtp\.fc\.": "model.layers.fc.",
```
Comment (govind-ramnarayan) [inline tensorrt_llm/_torch/auto_deploy/models/custom/modeling_eagle.py:152]: Wait, is there really an "fc" in the weights? Do we just silently drop this due to this code (which only initializes the "fc" module if the number of capture layers is more than 1)? https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/auto_deploy/models/custom/modeling_eagle.py#L526-L535

In this case, we need some sort of unit test to make sure we load all the weights and are missing nothing in our module definition...

File: tensorrt_llm/_torch/auto_deploy/models/eagle.py [hunk 5]
```diff
@@ -62,6 +62,9 @@ def _build_model(self, device: DeviceLikeType) -> nn.Module:
 
         # Get model type for config
         model_type = model_config.model_type
+        if model_type == "qwen3_5_moe" and hasattr(model_config, "text_config"):
```
Comment (govind-ramnarayan) [inline tensorrt_llm/_torch/auto_deploy/models/eagle.py]: Can you clarify this with a comment? It's not good to have model-specific code in this factory file. I would prefer putting something like this into the modeling code file and maybe calling a utility if we need. But first please clarify and also reply to this comment to explain to me.

File: examples/auto_deploy/model_registry/configs/qwen3.5_moe_400b_mtp.yaml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+cuda_graph_config:
+  max_batch_size: 32
+  batch_sizes: [1, 2, 3, 4, 5, 6, 7, 8, 16, 32]
+kv_cache_config:
+  enable_block_reuse: false
+  free_gpu_memory_fraction: 0.7
+  tokens_per_block: 32
+speculative_config:
+  decoding_type: MTP
+  max_draft_len: 1
```
Comment (govind-ramnarayan) [inline examples/auto_deploy/model_registry/configs/qwen3.5_moe_400b_mtp.yaml]: Let's try increasing this draft_len. Is there any test for Qwen3Next in PyTorch backend that we can use for inspiration? Otherwise can try... 6?

File: examples/auto_deploy/model_registry/configs/qwen3.5_moe_400b_mtp.yaml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+# Keep this as an overlay on top of qwen3.5_moe_400b.yaml so the base target
+# model remains the source of truth for sharding and quantization settings.
+#
+# MTP adds draft-layer KV plus managed recurrent GDN state and intermediate
+# speculative state. Use a smaller serving shape than the non-MTP baseline so
+# cache initialization leaves room for distributed workspaces and CUDA graph
+# capture.
+enable_chunked_prefill: false
+max_num_tokens: 8192
+max_batch_size: 32
```
Comment (govind-ramnarayan) [inline examples/auto_deploy/model_registry/configs/qwen3.5_moe_400b_mtp.yaml]: Note: May need to reduce the batch size to fit on H100s when we increase speculative length. Prioritize increasing speculative length as long as we can fit it in batch_size=8 or greater.

File: tensorrt_llm/_torch/auto_deploy/models/custom/modeling_eagle.py [hunk 8]
```diff
@@ -74,10 +75,12 @@ def get_eagle_layers(config, model_type: str) -> Union[nn.ModuleList, nn.Module]
             layers = build_llama_eagle_layers(config)
         case "nemotron_h":
             layers = build_nemotron_eagle_layers(config)
+        case "qwen3_5_moe" | "qwen3_5_moe_text":
```
Comment (govind-ramnarayan) [inline tensorrt_llm/_torch/auto_deploy/models/custom/modeling_eagle.py]: nit: What is the difference between "qwen3_5_moe" and "qwen3_5_moe_text". Can you walk me through the code path that creates this difference?

File: tensorrt_llm/_torch/auto_deploy/models/custom/modeling_qwen3_5_moe.py [hunk 9]
```diff
@@ -689,6 +689,64 @@ def forward(
...
+class Qwen3_5MoeEagleLayer(nn.Module):
+    """Qwen3.5 MTP layer for the AutoDeploy Eagle one-model path."""
+
+    def __init__(self, config: Qwen3_5MoeTextConfig, layer_idx: int):
+        super().__init__()
+        self.pre_fc_norm_embedding = Qwen3_5MoeRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+        self.pre_fc_norm_hidden = Qwen3_5MoeRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+        self.fc = nn.Linear(config.hidden_size * 2, config.hidden_size, bias=False)
+        self.self_attn = Qwen3_5MoeAttention(config, layer_idx)
+        self.mlp = Qwen3_5MoeSparseMoeBlock(config)
```
Comment (govind-ramnarayan) [inline tensorrt_llm/_torch/auto_deploy/models/custom/modeling_qwen3_5_moe.py:830]: a bit weird to call this MLP when it is an MoE. I guess this is needed for the checkpoint? Thoughts about changing it? I'm not sure if it is just customary to call it MLP when it is an MoE.

File: tensorrt_llm/_torch/auto_deploy/models/custom/modeling_qwen3_5_moe.py [hunk 10]
```diff
@@ -2937,7 +3015,5 @@ def init_input_processor(self, base):
 AutoConfig.register("qwen3_5_moe_text", Qwen3_5MoeTextConfig, exist_ok=True)
 
 AutoModelForCausalLMFactory.register_custom_model_cls("Qwen3_5MoeTextConfig", Qwen3_5MoeForCausalLM)
-AutoModelForCausalLMFactory.register_custom_model_cls(
-    "Qwen3_5MoeConfig", Qwen3_5MoeForConditionalGeneration
-)
+AutoModelForCausalLMFactory.register_custom_model_cls("Qwen3_5MoeConfig", Qwen3_5MoeForCausalLM)
```
Comment (govind-ramnarayan) [inline tensorrt_llm/_torch/auto_deploy/models/custom/modeling_qwen3_5_moe.py]: I see - so this is how we are redirecting Qwen3_5 MoE to the text model? Let's avoid this - let's try to just run our test with a model that has Qwen3_5MoeTextConfig. Maybe we can just overwrite the model type or something in our test for now?

File: tests/integration/defs/accuracy/test_llm_api_autodeploy.py [hunk 11]
```diff
@@ -1014,25 +1023,82 @@ def get_default_sampling_params(self):
...
+                          extra_evaluator_kwargs=self.EXTRA_EVALUATOR_KWARGS)
+
+    @skip_pre_hopper
+    @pytest.mark.skip_less_device_memory(80000)
+    @pytest.mark.parametrize("world_size", [8])
+    def test_fp8_mtp_gsm8k(self, world_size, mocker, monkeypatch):
+        if get_device_count() < world_size:
+            pytest.skip("Not enough devices for world size, skipping test")
+        kwargs = self.get_default_kwargs()
+        monkeypatch.setenv("TRTLLM_ACCURACY_NO_REFERENCE", "1")
```
Comment (govind-ramnarayan) [inline tests/integration/defs/accuracy/test_llm_api_autodeploy.py]: No need for this, just use this env var locally. Better yet - add in the appropriate accuracy for this test.

File: tests/integration/defs/accuracy/test_llm_api_autodeploy.py [hunk 12]
```diff
@@ -1014,25 +1023,82 @@ def get_default_sampling_params(self):
...
     @pytest.mark.parametrize("world_size", [8])
-    def test_bf16(self, world_size, mocker):
+    def test_fp8_gsm8k(self, world_size, mocker, monkeypatch):
         if get_device_count() < world_size:
             pytest.skip("Not enough devices for world size, skipping test")
         kwargs = self.get_default_kwargs()
-        sampling_params = self.get_default_sampling_params()
-        model_path = hf_id_to_local_model_dir(self.MODEL_NAME)
+        kwargs.update(self.get_fp8_gsm8k_runtime_overrides())
+        monkeypatch.setenv("TRTLLM_ACCURACY_NO_REFERENCE", "1")
```
Comment (govind-ramnarayan) [inline tests/integration/defs/accuracy/test_llm_api_autodeploy.py]: Ditto to below - no monkeypatch.

File: tests/integration/defs/accuracy/test_llm_api_autodeploy.py [hunk 13]
```diff
@@ -1014,25 +1023,82 @@ def get_default_sampling_params(self):
...
+                    "enabled": False
+                },
+            },
+        }
+
+    @skip_pre_hopper
     @pytest.mark.skip_less_device_memory(80000)
     @pytest.mark.parametrize("world_size", [8])
-    def test_bf16(self, world_size, mocker):
+    def test_fp8_gsm8k(self, world_size, mocker, monkeypatch):
```
Comment (govind-ramnarayan) [inline tests/integration/defs/accuracy/test_llm_api_autodeploy.py]: Let's reinstate the bf16 test - we just changed to fp8 for a quick comparison. Let's revert this test completely.

File: tests/integration/defs/accuracy/test_llm_api_autodeploy.py [hunk 14]
```diff
@@ -1014,25 +1023,82 @@ def get_default_sampling_params(self):
...
+            mocker.patch.object(GSM8K, "MAX_OUTPUT_LEN",
+                                self.GSM8K_MAX_OUTPUT_LEN)
+            task = GSM8K(self.MODEL_NAME)
+            task.evaluate(llm,
+                          extra_evaluator_kwargs=self.EXTRA_EVALUATOR_KWARGS)
+
+    @skip_pre_hopper
+    @pytest.mark.skip_less_device_memory(80000)
+    @pytest.mark.parametrize("world_size", [8])
+    def test_fp8_mtp_gsm8k(self, world_size, mocker, monkeypatch):
```
Comment (govind-ramnarayan) [inline tests/integration/defs/accuracy/test_llm_api_autodeploy.py]: Since we are doing bf16 tests for no MTP, maybe let's try just doing BF16 + MTP and see if we have enough memory for that - otherwise we can make this test FP8 and possibly make the no-MTP test FP8 for comparison so that we can see that MTP does not decrease accuracy.

File: tests/unittest/auto_deploy/singlegpu/custom_ops/fla/test_fla_cached_gated_delta_rule.py [hunk 15]
```diff
@@ -84,6 +73,73 @@ def _preprocess_for_reference(q, k, a, b_proj, A_log, dt_bias, num_v_heads=None)
     return q_norm, k_norm, g, beta
 
 
+def _make_extend_case(device, dtype, *, num_k_heads=2, num_v_heads=2):
```
Comment (govind-ramnarayan) [inline tests/unittest/auto_deploy/singlegpu/custom_ops/fla/test_fla_cached_gated_delta_rule.py]: This name is super vague - let's name it "make_extend_kernel_inputs()"

File: tests/unittest/auto_deploy/singlegpu/custom_ops/fla/test_fla_cached_gated_delta_rule.py [hunk 16]
```diff
@@ -287,6 +345,107 @@ def test_prefill_only(gdr_env, num_k_heads, num_v_heads):
...
+            beta=beta[:, start:end],
+            scale=scale,
+            initial_state=initial_state,
+            output_final_state=True,
+            use_qk_l2norm_in_kernel=False,
+        )
+
+        y_ref[:, start:end] = y_seq.to(dtype)
+        for step_idx in range(tokens_per_extend):
+            _, prefix_final_state = fused_recurrent_gated_delta_rule_fwd(
```
Comment (govind-ramnarayan) [inline tests/unittest/auto_deploy/singlegpu/custom_ops/fla/test_fla_cached_gated_delta_rule.py]: It makes more sense here to use the prefill kernel for each step, use the extend kernel for the "maximum" step size, and then compare the states after prefill kernels to the extend kernel. I don't think there's any reason to dig into the kernels like this either. Just do something like:

- Set up prefill inputs. For i=1, ... num_extend_tokens, run a prefill with i tokens and collect the state afterwards.
- Set up extend inputs. Run a single extend request for num_extend_tokens.

Do both of these directly with the autodeploy custom op - not the individual kernel calls. 

We can just set this up so num_extend_tokens is something small - like 2 or 3. And the sizes should be small enough that this test runs *fast*.

File: tests/unittest/auto_deploy/singlegpu/models/test_qwen3_5_moe_mtp.py [hunk 17]
```diff
@@ -0,0 +1,579 @@
...
+    TargetModelExportInfo,
+)
+from tensorrt_llm._torch.auto_deploy.models.hf import AutoModelForCausalLMFactory
+from tensorrt_llm._torch.auto_deploy.transform.interface import TransformConfig
+from tensorrt_llm._torch.auto_deploy.transform.library.hidden_states import (
+    DetectHiddenStatesForCapture,
+)
+from tensorrt_llm.llmapi import MTPDecodingConfig
+
+
```
Comment (govind-ramnarayan) [inline tests/unittest/auto_deploy/singlegpu/models/test_qwen3_5_moe_mtp.py]: Note: Throughout this file remove "qwen_3" strings in the test titles - it's not needed in this file. 

Let's actually append these tests to the existing qwen3_5_moe modeling tests, since they are included in that file. We can put them at the end and separate from the others with a note that they are for testing MTP modeling definitions and utilities.
